service/snap/snap.go

```diff
@@ -0,0 +1,422 @@
+// Copyright 2019 Canonical Ltd.
+// Licensed under the AGPLv3, see LICENCE file for details.
+
+// Package snap is a minimal service.Service implementation modelled on service/upstart.
+package snap
+
+import (
+	"bytes"
+	"fmt"
+	"os/exec"
+	"regexp"
+	"runtime"
+	"strings"
+
+	"github.com/juju/errors"
+	"github.com/juju/loggo"
+	"github.com/juju/utils/featureflag"
+	"github.com/juju/utils/set"
+	"github.com/juju/utils/shell"
+
+	"github.com/juju/juju/feature"
+	"github.com/juju/juju/service/common"
+)
+
+const (
+	Command           = "snap"
+	CustomConfigPath  = "/var/snap/juju-db/common/custom.config"
+	DefaultConfigPath = "/var/snap/juju-db/common/default.config"
+	DefaultDataPath   = "/var/snap/juju-db/common"
+)
+
+var (
+	logger = loggo.GetLogger("juju.service.snap")
+
+	// snapNameRe is derived from https://github.com/snapcore/snapcraft/blob/a2ef08109d86259a0748446f41bce5205d00a922/schema/snapcraft.yaml#L81-106
+	// but does not test for "--"
+	snapNameRe = regexp.MustCompile("^[a-z0-9][a-z0-9-]{0,39}[^-]$")
+
+	modes    = set.NewStrings("devmode", "classic", "jailmode")
+	channels = set.NewStrings("edge", "beta", "candidate", "stable")
+)
+
+// IsRunning indicates whether Snap is currently running on the system.
+// When the snap command (normally installed to /usr/bin/snap) cannot be
+// detected, IsRunning returns (false, nil). Other errors result in (false, err).
+func IsRunning() (bool, error) {
+	if runtime.GOOS == "windows" {
+		return false, nil
+	}
+
+	cmd := exec.Command(Command, "version")
+	out, err := cmd.CombinedOutput()
+	logger.Debugf("snap version output: %#v", string(out[:]))
+	if err == nil {
+		if !featureflag.Enabled(feature.MongoDbSnap) {
```

Comment:
No need to use feature flag at this level.
The primitives in this package should be purely snap related